app/controllers/ticket.js

```diff
@@ -0,0 +1,430 @@
+  },
+
+  deleteTicket: async (req, res, next) => {
+    const { id } = req.params
+    const userId = req.user.id.toString()
+    try {
+      const ticket = await TicketModel.findById(id)
+      if (!ticket) {
+        return res.status(HttpStatus.NOT_FOUND).json({ error: 'No ticket exist' })
+      }
+      if (userId !== ticket.createdBy && !req.user.isAdmin && !req.user.isTicketsModerator) {
+        // Only user who created the ticket and admin can delete the ticket
+        return res.status(HttpStatus.FORBIDDEN).json({ error: 'Bad delete request' })
+      }
+      await TicketModel.findByIdAndRemove(id)
+      res.status(HttpStatus.OK).json({ ticket: ticket })
+    } catch (error) {
+      HANDLER.handleError(res, {
+        code: error.code || HttpStatus.BAD_REQUEST,
+        ...error
+      })
+    }
+  },
+
+  editTag: async (req, res, next) => {
+    const { id } = req.params
+    const { tags } = req.body // tags is the array of tags to add
+    const userId = req.user.id.toString()
+    try {
+      const ticket = await TicketModel.findById(id)
+      if (!ticket) {
+        return res.status(HttpStatus.NOT_FOUND).json({ error: 'No ticket exist' })
+      }
+      if (userId !== ticket.createdBy && !req.user.isAdmin && !req.user.isTicketsModerator) {
+        // Only user who created the ticket and admin can edit ticket tags
+        return res.status(HttpStatus.FORBIDDEN).json({ error: 'Edit Forbidden by user' })
+      }
+      ticket.tags = [...new Set(tags)]
+      await ticket.save()
+      res.status(HttpStatus.OK).json({ ticket: ticket })
+    } catch (error) {
+      console.log(error)
+      HANDLER.handleError(res, {
+        code: error.code || HttpStatus.BAD_REQUEST,
+        ...error
+      })
+    }
+  },
+
+  addTag: async (req, res, next) => {
+    const { id, tag } = req.params
+    const userId = req.user.id.toString()
+    try {
+      const ticket = await TicketModel.findById(id)
+      if (!ticket) {
+        return res.status(HttpStatus.NOT_FOUND).json({ error: 'No ticket exist' })
+      }
+      if (userId !== ticket.createdBy && !req.user.isAdmin && !req.user.isTicketsModerator) {
+        // Only user who created the ticket and admin can add tag to the ticket
+        return res.status(HttpStatus.FORBIDDEN).json({ error: 'Edit Forbidden by user' })
+      }
+      ticket.tags.addToSet(tag)
+      await ticket.save()
+      res.status(HttpStatus.OK).json({ ticket: ticket })
+    } catch (error) {
+      console.log(error)
+      HANDLER.handleError(res, {
+        code: error.code || HttpStatus.BAD_REQUEST,
+        ...error
+      })
+    }
+  },
+
+  // Create Comment of a Ticket
+  createComment: async (req, res, next) => {
+    const { id } = req.params
+    const userId = req.user.id.toString()
+    try {
+      const ticket = await TicketModel.findById(id)
+      if (!ticket) {
+        return res.status(HttpStatus.NOT_FOUND).json({ error: 'No ticket exist' })
+      }
+      ticket.comments.push({
```

Comment:
latest comment should be on top, use unshift

Reply:
Yes, that makes sense without context to the frontend, but the frontend shows the comments like a thread for the oldest to latest.